core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1967,39 +1967,138 @@ private[spark] object Utils extends Logging {
+        val (service, port) = startOnce(startPort)
+
+        if (port != -1) {
+          (service.get, port)
+        } else {
+          // try other
+          startFromAvailable()
+        }
+      } else {
+        // try random
+        startFromAvailable(true)
+      }
+
+    } else {
+
+      val maxRetries = portMaxRetries(conf)
+
       // Do not increment port if startPort is 0, which is treated as a special port
-      val tryPort = if (startPort == 0) {
+      val tryPort = (x: Int) => if (startPort == 0) {
         startPort
       } else {
         // If the new port wraps around, do not try a privilege port
-        ((startPort + offset - 1024) % (65536 - 1024)) + 1024
-      }
-      try {
-        val (service, port) = startService(tryPort)
-        logInfo(s"Successfully started service$serviceString on port $port.")
-        return (service, port)
-      } catch {
-        case e: Exception if isBindCollision(e) =>
-          if (offset >= maxRetries) {
-            val exceptionMessage =
-              s"${e.getMessage}: Service$serviceString failed after $maxRetries retries!"
-            val exception = new BindException(exceptionMessage)
-            // restore original stack trace
-            exception.setStackTrace(e.getStackTrace)
-            throw exception
-          }
-          logWarning(s"Service$serviceString could not bind on port $tryPort. " +
-            s"Attempting port ${tryPort + 1}.")
+        ((startPort + x - 1024) % (65536 - 1024)) + 1024
       }
+
+      retryPort(tryPort, maxRetries)
     }
-    // Should never happen
-    throw new SparkException(s"Failed to start service$serviceString on port $startPort")
+  }
+
+  /**
+   * Checks if there is any port restriction (currently used only with mesos)
+   * @param conf the SparkConfig to use
+   * @return true if port range restriction holds false otherwise
+   */
+  def hasPortRangeRestriction(conf: SparkConf): Boolean = {
+    // Caution: the property is for internal use only
+    conf.getOption("spark.mesos.executor.port.ranges").isDefined
+  }
+
+  /**
+   * Gets the allowed port ranges to use (currently  used only with mesos)
+   * @param conf the SparkConfig to use
+   * @return the string containing a list of port ranges in the format:
+   * (begin_port_number_1, end_port_number_1):...:(begin_port_number_n, end_port_number_n)
+   */
+  def getPortRangeRestriction(conf: SparkConf): Option[String] = {
+    // Caution:  the property is for internal use only
+    conf.getOption("spark.mesos.executor.port.ranges")
+  }
+
+  /**
+   * Sets the allowed port ranges to use (currently used only with mesos).
+   * @param conf the SparkConfig to use
+   * @param ports a list of port ranges with format: (begin_port_number, end_port_number)
+   */
+  def setPortRangeRestriction(conf: SparkConf, ports: List[(Long, Long)]): Unit = {
+    // Caution: the property is for internal use only
+    conf.set("spark.mesos.executor.port.ranges", ports.mkString(":"))
```

Comment:
this is kind of an anti-pattern; `conf` is supposed to be immutable once we create the `SparkContext`. If we set things after that some things might get the new value but others might not. It's a confusing pattern that we try to shy away from in general.